src/main/java/org/radarbase/appserver/service/RadarProjectService.java

```diff
@@ -1,15 +1,135 @@
+import org.radarbase.appserver.exception.InvalidProjectDetailsException;
+import org.radarbase.appserver.exception.NotFoundException;
+import org.radarbase.appserver.repository.ProjectRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.*;
 
+/**
+ * {@link Service} for interacting with the {@link Project} {@link javax.persistence.Entity} using the
+ * {@link ProjectRepository}.
+ *
+ * @author yatharthranjan
+ */
+@Service
 public class RadarProjectService {
 
+    @Autowired
+    private ProjectRepository projectRepository;
+
+    @Transactional(readOnly = true)
     public RadarProjects getAllProjects() {
-        return null;
+        RadarProjects projects = new RadarProjects();
+        return new RadarProjects().setProjects(
+                ConverterFactory.getProjectConverter().entitiesToDtos(projectRepository.findAll()));
+    }
+
+    @Transactional(readOnly = true)
+    public RadarProjectDto getProjectById(Long id) {
+        Optional<Project> project = projectRepository.findById(id);
+
+        if (project.isPresent()) {
+            return ConverterFactory.getProjectConverter().entityToDto(project.get());
+        } else {
+            throw new NotFoundException("Project not found with id" + id);
+        }
     }
 
-    public RadarProjectDto getProjectById(String id) {
-        return null;
+    @Transactional(readOnly = true)
+    public RadarProjectDto getProjectByProjectId(String projectId) {
+        Optional<Project> project = projectRepository.findByProjectId(projectId);
+
+        if (project.isPresent()) {
+            return ConverterFactory.getProjectConverter().entityToDto(project.get());
+        } else {
+            throw new NotFoundException("Project not found with projectId" + projectId);
+        }
+    }
+
+    @Transactional(readOnly = true)
+    public Project getProjectEntityByProjectId(String projectId) {
+        Optional<Project> project = projectRepository.findByProjectId(projectId);
+
+        if (project.isPresent()) {
+            return project.get();
+        } else {
+            throw new NotFoundException("Project not found with projectId" + projectId);
+        }
+    }
+
+    @Transactional
+    public RadarProjectDto addProject(RadarProjectDto projectDto) {
+
+        Optional<Project> project;
+        if (projectDto.getProjectId() != null && !projectDto.getProjectId().isEmpty()) {
+            project = projectRepository.findByProjectId(projectDto.getProjectId());
+        } else if (projectDto.getId() != null) {
+            project = projectRepository.findById(projectDto.getId());
+        } else {
+            throw new InvalidProjectDetailsException(projectDto, new IllegalArgumentException("At least one of 'id' or 'project id' must be supplied"));
+        }
+
+        Project resultProject;
+        if(project.isPresent()) {
```

Comment:
It looks a bit too complicated to me.This looks like a hybrid of add and update. Is this intentional? 
If I want to add a project, 
- I will not add something which has an Id and throw `IllegalArgumentException`.
- If there is no `projectId`, then throw illegal argument.
- If there is a projectId query the object, if it exist then throw ConflictingRequest since one already exist.  
- if not create. 
This is what I would ideally expect from an `add` method.

Reply:
Sorry yes. I think I forgot to update this one. Its from an old piece of code when I just started working on this and never got back to fixing this (but did fix in other services).